components/sidebar/sidebar_channel/sidebar_channel.tsx

```diff
@@ -0,0 +1,134 @@
+     */
+    currentTeamName: string;
+
+    /**
+     * Number of unread mentions in this channel
+     */
+    unreadMentions: number;
+
+    /**
+     * Number of unread messages in this channel
+     */
+    unreadMsgs: number;
+
+    /**
+     * User preference of whether the channel can be marked unread
+     */
+    showUnreadForMsgs: boolean;
+
+    /**
+     * Gets the ref for a given channel id
+     */
+    getChannelRef: (channelId: string) => HTMLLIElement | undefined;
+
+    /**
+     * Sets the ref for the sidebar channel div element, so that it can be used by parent components
+     */
+    setChannelRef: (channelId: string, ref: HTMLLIElement) => void;
+
+    /**
+     * If category is collapsed
+     */
+    isCategoryCollapsed: boolean;
+
+    /**
+     * Is the channel the currently focused channel
+     */
+    isCurrentChannel: boolean;
+};
+
+type State = {
+
+};
+
+export default class SidebarChannel extends React.PureComponent<Props, State> {
+    isUnread = () => {
+        return this.props.unreadMentions > 0 || (this.props.unreadMsgs > 0 && this.props.showUnreadForMsgs);
+    }
+
+    isCollapsed = (props: Props) => {
+        return props.isCategoryCollapsed && !this.isUnread() && !props.isCurrentChannel;
+    }
+
+    componentDidUpdate(prevProps: Props) {
+        if (this.isCollapsed(this.props) !== this.isCollapsed(prevProps)) {
+            const channelElement = this.getRef();
+            if (channelElement) {
+                channelElement.classList.add('animating');
+            }
+        }
+    }
+
+    removeAnimation = () => {
+        const channelElement = this.getRef();
+        if (channelElement) {
+            channelElement.classList.remove('animating');
+        }
+    }
+
+    getRef = () => {
+        return this.props.getChannelRef(this.props.channel.id);
+    }
+
+    setRef = (ref: HTMLLIElement) => {
+        this.props.setChannelRef(this.props.channel.id, ref);
+    }
+
+    render() {
+        const {channel, currentTeamName} = this.props;
+
+        let ChannelComponent: React.ComponentType<{channel: Channel; currentTeamName: string}> = SidebarBaseChannel;
+        if (channel.type === Constants.DM_CHANNEL) {
+            ChannelComponent = SidebarDirectChannel;
+        } else if (channel.type === Constants.GM_CHANNEL) {
+            ChannelComponent = SidebarGroupChannel;
+        }
+
+        return (
+            <li
+                role='listitem'
+                draggable='false'
+                ref={this.setRef}
+                className={classNames('SidebarChannel', {
+                    collapsed: this.isCollapsed(this.props),
+                    unread: this.isUnread(),
+                    active: this.props.isCurrentChannel,
+                })}
+                onTransitionEnd={this.removeAnimation}
+            >
+                <ChannelComponent
+                    channel={channel}
+                    currentTeamName={currentTeamName}
```

Comment:
This could be left for a future improvement, but do we really need to pass this around everywhere? Those components are already connected to the store, so they can get it themselves. The previous code likely passed this around because it was rewritten in the period where we wanted less components connecting to the flux stores.

Reply:
It's more about reducing the amount of times the store has to actually fetch it. I'm all for having redux grab it instead, but I'm not sure what's more performant. Let me know.

Reply:
Yeah, I'm not sure what's better. We've kind of gone back in forth on it since we don't know the overhead of `connect`. Best to leave it as-is for now

Reply:
It's fine, Mostly a micro optimisation and not something we need to worry about IMO.  As long as we are not passing them between components which are not consuming it is fine.